### ISSUE TITLE
mysql5.7 Connection issues

### ISSUE THREAD

**Issue by Ericj100** [theangryangel/logstash-output-jdbc#48]
My Environment

* Version of plugin used: 0.2.10
* Version of Logstash used:  2.2.1
* Version of Mysql used:  5.7.13
* Database engine & version you're connecting to: 

> com.mysql.jdbc.Driver & mysql-connector-java-5.1.39

* Have you checked you've met the Logstash requirements for Java versions?:

>  yes,java version "1.8.0_45"

Using Mysql with the JAR in vendor/jar/jdbc with the jar in your example and basically the same connection string. 
```
output {
    jdbc {
        driver_class => "com.mysql.jdbc.Driver"
        connection_string => "jdbc:mysql://172.16.66.184:3306/jsontest?user=root&password=Cmcc_123&useSSL=false"
        #connection_timeout => 5000
        statement => [ "replace into user (id,name, lastlogininfo) values (10,?,?)", "%{name}", "%{lastlogininfo}" ]
    }
}
```
I am getting this output in my logstash.log:

> The error reported is: 
>   Exception during pool initialization: HikariPool-30 - Connection is not available, request timed out after 10000ms.

With the same connection string of configuration, when I connection to mysql 5.6, it is fine, but when I connection to mysql 5.7, it is error and the log information as above

And I have try the JDBC connection with a simple Java class instead:
```
HikariDataSource ds = new HikariDataSource();

        ds.setJdbcUrl("jdbc:mysql://172.16.66.184:3306/jsontest?useSSL=false");
        ds.setUsername("root");
        ds.setPassword("Cmcc_123");
        ds.setDriverClassName(com.mysql.jdbc.Driver.class.getName());
        ds.setAutoCommit(false);
```
It is fine too.

Kind of stuck on this, any help would be appreciated..

**Comment by theangryangel** [comment]
What version of Hikari did you use to test?

**Comment by Ericj100** [comment]
in /vendor/bundle/jruby/1.9/gems/logstash-output-jdbc-0.2.10/lib/com/zaxxer/HikariCP path, logstash-output-jdbc-0.2.10 self have  two version:

> 2.4.2 and 2.4.6

> ./vendor/bundle/jruby/1.9/gems/logstash-output-jdbc-0.2.10/lib/com/zaxxer/HikariCP/2.4.2/HikariCP-2.4.2.jar
> ./vendor/bundle/jruby/1.9/gems/logstash-output-jdbc-0.2.10/lib/com/zaxxer/HikariCP/2.4.6/HikariCP-2.4.6.jar

I'm a little dizzy, i don't know why to this

And I have try the JDBC connection with a simple Java class, I have try HikariCP with tow version:

> 2.4.2 and 2.4.6
they are fine both

**Comment by theangryangel** [comment]
Hmmm. That shouldnt have both versions in the gem - it's probably from when I was switching between the v2 and v5 branch and I didn't do a cleanup before release - oops :/ 

Does the error get logged when Logstash starts? Or after it's been running for a while? I just need to know if it's the first connection test, or if its when it tries to get a connection from the pool later (in which case it's a connection leak).

**Comment by Ericj100** [comment]
I have remove "...HikariCP/2.4.2..." directory, the error is occured too.

when Logstash starts, and the timeout is reached, the error is occured. So the error get logged when Logstash starts, and these is no connection to this. Only the initialization of HikariCP occur error

**Comment by theangryangel** [comment]
Can you try this config @Ericj100 

```
output {
    jdbc {
        driver_class => "com.mysql.jdbc.Driver"
        connection_string => "jdbc:mysql://172.16.66.184:3306/jsontest?useSSL=false"
        username => "root"
        password => "Cmcc_123"
        connection_timeout => 20000
        statement => [ "replace into user (id,name, lastlogininfo) values (10,?,?)", "%{name}", "%{lastlogininfo}" ]
    }
}
```

If that doesn't work, let me know and I'll power up a virtual machine tonight (if I have time) and I'll see what I can figure out.

**Comment by Ericj100** [comment]
Hi, @theangryangel 
    Thank you for your reply.
    I have tried that config, but is not ok too.
    But when I add a parameter "max_pool_size" as follow
```
output {
    jdbc {
        driver_class => "com.mysql.jdbc.Driver"
        connection_string => "jdbc:mysql://172.16.66.184:3306/jsontest?useSSL=false"
        username => "root"
        password => "Cmcc_123"
        connection_timeout => 20000
        max_pool_size => 1
        statement => [ "replace into user (id,name, lastlogininfo) values (10,?,?)", "%{name}", "%{lastlogininfo}" ]
    }
}
```
    It is fine, although I don't quite understand. 
    And i have test with this parameter. I found some case:
    1. max_pool_size <= 3, it is fine; 
    2. 3 < max_pool_size <= 6, it will be happened error with HikariPool Connection is not available;
    3. max_pool_size > 6, it will be heppened error as follow:
```
DEBUG - HikariPool-21 - Cannot acquire connection from data source
com.mysql.jdbc.exceptions.jdbc4.MySQLNonTransientConnectionException: Data source rejected establishment of connection,  message from server: "Too many connections"
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
▽       at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
        at com.mysql.jdbc.Util.getInstance(Util.java:387)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:917)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:896)
        at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:885)
        at com.mysql.jdbc.MysqlIO.doHandshake(MysqlIO.java:1042)
        at com.mysql.jdbc.ConnectionImpl.coreConnect(ConnectionImpl.java:2253)
        at com.mysql.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:2284)
        at com.mysql.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:2083)
        at com.mysql.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:806)
        at com.mysql.jdbc.JDBC4Connection.<init>(JDBC4Connection.java:47)
        at sun.reflect.GeneratedConstructorAccessor14.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
        at com.mysql.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:410)
        at com.mysql.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:328)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:95)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:101)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:314)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:171)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:441)
        at com.zaxxer.hikari.pool.HikariPool.access$300(HikariPool.java:66)
        at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:576)
        at com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:569)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```
        I have found logstash Default pipeline workers is 24, so we have 24 Hikari connections.
       That max_pool_size parameter, I still have some not too clear of its role

**Comment by theangryangel** [comment]
So HikariCP is a connection pool. The idea is that it opens X connections and then when the plugin needs one it just asks for one. The idea is that rather than making a connection every single time it needs to push a set of events, it just keeps connections open, but idle (or rather HikariCP does). It also deals with reconnections, etc.

Before Logstash 2.2 you'd only get 1 instance of a plugin. [In 2.2 elastic.co changed this](https://www.elastic.co/blog/logstash-2-2-0-and-2-1-2-released), and how events get processed. The most important thing is that unless you declare the plugin as threadsafe, you get multiple instances of the plugin. 

The side effect of this plugin not being marked as threadsafe (yet) is that Logstash now creates multiple instances of this plugin. So the total possible number of connections that this plugin can open, right now, would be number of pipelines * max_pool_size connections. This is why you're getting so many connections and why it's failing (basically maxing out the connection limit).

I'll be 100% honest, I knew that the Logstash architecture had changed, but I hadn't realised that it was spawning so many instances of the plugin under the new pipeline. Frankly I didn't exactly expect something that major changing in a minor version, and I've only been trying to keep up with Logstash development since they released the v5 alpha some months ago. 

**TL;DR**: The real fix is going to have to be make this plugin threadsafe, and that'll break compatibility with 2.0 and 2.1. Probably. Guess it's going to have to be done though as opening that many connections is just too much.

I'll start work on that after work tonight. I probably won't get time during my day job unfortunately :(

**Comment by Ericj100** [comment]
It is great. Thank you for your explanation. I have understood now :(
 My clumsy English, unexpectedly you can understand haha

**Comment by theangryangel** [comment]
Your english is good Eric - no need to apologise :)

**Comment by theangryangel** [comment]
@Ericj100 I've pushed 0.3.0 to rubygems. This release marks the plugin as threadsafe, so only 1 instance gets created, which will prevent this problem from happening in future.

If you do update, I'd recommend removing max_pool_size or altering the max_pool_size to match the number of workers :)

**Comment by Ericj100** [comment]
OK, I've knowned it. And I will try it, thanks @theangryangel  :)
